From c0539d08cb0ca8d1964151a6e23875afef633e75 Mon Sep 17 00:00:00 2001
From: "Mazlan, Hazwan Arif" <hazwan.arif.mazlan@intel.com>
Date: Mon, 3 Jul 2023 07:21:45 +0800
Subject: [PATCH 27/30] ui/gtk: enable ungrabbing fullscreened widget

this change makes the hotkey working in fullscreen mode
and can let user switch VM in fullscreen mode

v2: s->grab_item wasn't initialized by the time it was
    referenced. So fixed it.

Signed-off-by: Kim, Dongwon <dongwon.kim@intel.com>
Signed-off-by: He, Xiao <xiao.he@intel.com>
Signed-off-by: Mazlan, Hazwan Arif <hazwan.arif.mazlan@intel.com>
---
 ui/gtk.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/ui/gtk.c b/ui/gtk.c
index 4a069ea9e5..b84ce9644e 100644
--- a/ui/gtk.c
+++ b/ui/gtk.c
@@ -2216,6 +2216,12 @@ static void gd_menu_grab_input(GtkMenuItem *item, void *opaque)
     gd_update_cursor(vc);
 }
 
+static void gd_accel_grab_input(void *opaque)
+{
+    GtkDisplayState *s = opaque;
+    gtk_menu_item_activate(GTK_MENU_ITEM(s->grab_item));
+}
+
 static void gd_change_page(GtkNotebook *nb, gpointer arg1, guint arg2,
                            gpointer data)
 {
@@ -2844,8 +2850,6 @@ static GtkWidget *gd_create_menu_view(GtkDisplayState *s, DisplayOptions *opts)
     s->zoom_out_item = gtk_menu_item_new_with_mnemonic(_("Zoom _Out"));
     gtk_menu_item_set_accel_path(GTK_MENU_ITEM(s->zoom_out_item),
                                  "<QEMU>/View/Zoom Out");
-    gtk_accel_map_add_entry("<QEMU>/View/Zoom Out", GDK_KEY_minus,
-                            HOTKEY_MODIFIERS);
     gtk_menu_shell_append(GTK_MENU_SHELL(view_menu), s->zoom_out_item);
 
     s->zoom_fixed_item = gtk_menu_item_new_with_mnemonic(_("Best _Fit"));
@@ -2871,6 +2875,11 @@ static GtkWidget *gd_create_menu_view(GtkDisplayState *s, DisplayOptions *opts)
         gtk_accel_map_add_entry("<QEMU>/View/Grab Input", GDK_KEY_g,
                                 HOTKEY_MODIFIERS);
         gtk_menu_shell_append(GTK_MENU_SHELL(view_menu), s->grab_item);
+        gtk_accel_group_connect(s->accel_group, GDK_KEY_g, HOTKEY_MODIFIERS, 0,
+                          g_cclosure_new_swap(G_CALLBACK(gd_accel_grab_input), s, NULL));
+        gtk_accel_label_set_accel(
+                          GTK_ACCEL_LABEL(gtk_bin_get_child(GTK_BIN(s->grab_item))),
+                          GDK_KEY_g, HOTKEY_MODIFIERS);
     }
 
     separator = gtk_separator_menu_item_new();
-- 
2.34.1

